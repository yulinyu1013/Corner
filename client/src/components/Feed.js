import React, { useState } from 'react';
import { Menu, Dropdown } from 'antd';
// import { useParams } from 'react-router';
import 'antd/dist/antd.css';
import option from '../images/more_vertical.svg';
import '../styles/feed.css';
import logo from '../images/logo.png';
import Comment from './Comment';
import { sendNotificationsforFlagPost } from '../fetchers/notificationApis';
import { updatePost } from '../fetchers/feedsApi';

function Feed({
  post,
  comments,
  deletePostHandler,
  setAllComments,
  allComments,
  allPosts,
  setAllPosts,
  cornerid,
  allUsers,
  admins,
  isLoggedIn,
  userName,
  avatar,
  userId,
  isAdmin,
}) {
  // all values below should get from localstorage or params
  // let { cornerId } = useParams();
  const cornerId = cornerid;
  // const userId = '1'; // get from localstorage
  // const userName = 'test'; // get from localstorage
  // const isAdmin = false; // get from localstroage
  const [flagged, setFlgged] = useState(post.isFlagged === 1);

  const flagPost = () => {
    if (!flagged) {
      setFlgged(true);
      const receivers = admins.join(',');
      sendNotificationsforFlagPost(userName, userId, receivers, cornerId, post.id);
      const flag = { isFlagged: 1 };
      updatePost(post.id, flag);
    }
  };

  const hidePost = () => {
    const newPosts = allPosts.map((p) => {
      if (p.id === post.id) {
        if (p.isHidden === 0) {
          p.isHidden = 1;
          const hidden = { isHidden: 1 };
          updatePost(post.id, hidden);
        } else {
          p.isHidden = 0;
          const hidden = { isHidden: 0 };
          updatePost(post.id, hidden);
        }
      }
      return p;
    });
    setAllPosts(newPosts);
  };

  const menu = (
    <Menu style={{ borderRadius: '7px', width: 100 }}>
      <Menu.Item key={1}>
        <div className="dropdown-item" role="presentation" onClick={() => { flagPost(); }}>{flagged ? 'Flagged' : 'Flag'}</div>
      </Menu.Item>
      {(post.creator === userId || isAdmin) && (
        <Menu.Item key={2}>
          <div className="dropdown-item" role="presentation" onClick={() => { hidePost(); }}>{post.isHidden === 0 ? 'Hide' : 'Unhide'}</div>
        </Menu.Item>
      )}
      {(post.creator === userId || isAdmin) && (
        <Menu.Item key={3}>
          <div className="dropdown-item" role="presentation" onClick={() => { deletePostHandler(post.id); }}>Delete</div>
        </Menu.Item>
      )}
    </Menu>
  );
  const dateTime = new Date(post.createdAt).toLocaleDateString();
  const [showMorePics, setShowMorePics] = useState(false);

  const showMorePicsHandler = () => {
    setShowMorePics(true);
  };

  return (
    <div className="feed-content">
      <div className="user-info">
        <div className="left-info">
          <img className="user-pic" src={post.avatar ? post.avatar : logo} alt="user-pic" />
          <div className="title">
            <p className="user-name">{`${post.creatorName} in ${post.cornerName}`}</p>
            {post.isHidden === 1 && <p className="user-hidden">The post is hidden</p>}
          </div>
        </div>
        <div className="right-info">
          <p className="time">{dateTime}</p>
          {cornerId
            && (
            <Dropdown overlay={menu} placement="bottomRight" trigger={['click']}>
              <img className="option" src={option} alt="option" />
            </Dropdown>
            )}
        </div>
      </div>
      <div className="post-content">
        <p className="text-content">{post.text}</p>
        {post.pic && post.pic.length > 0 && (
          <div className="feed-all-image">
            {post.pic.slice(0, 2).map((image) => (
              <img key={image} src={typeof image === 'string' ? image : Object.values(image)} alt="uploaded-pics" />
            ))}
            {!showMorePics && post.pic.length > 2 && (
              <div className="show-more-block" style={{ backgroundImage: `url(${Object.values(post.pic[2])})` }} role="presentation" onClick={() => showMorePicsHandler()}>click to show more</div>
            )}
            {showMorePics && (
              post.pic.slice(2).map((image) => (
                <img key={image} src={typeof image === 'string' ? image : Object.values(image)} alt="uploaded-pics" />
              ))
            )}
          </div>
        )}
        {post.video && (
          <div className="feed-video">
            <video src={post.video} controls><track kind="captions" /></video>
          </div>
        )}
        {post.audio && (
          <div className="feed-audio">
            <audio src={typeof post.audio === 'string' ? post.audio : Object.values(post.audio)} controls><track kind="captions" /></audio>
          </div>
        )}
      </div>
      <div className="reaction">
        {/* <img className="like-icon" src={like} alt="likes" />
        <p>5</p> */}
        <p>Â·</p>
        <p>{`${comments.length} comments`}</p>
      </div>
      <div className="comment-part">
        <Comment
          cornerId={post.corner}
          comment={comments}
          pid={post.id}
          setAllComments={setAllComments}
          allComments={allComments}
          isfromCornerPage={cornerId}
          allUsers={allUsers}
          cornerName={post.cornerName}
          isLoggedIn={isLoggedIn}
          userName={userName}
          avatar={avatar}
          userId={userId}
        />
      </div>
    </div>
  );
}

export default Feed;
