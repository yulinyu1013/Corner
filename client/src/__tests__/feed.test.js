import React from 'react';
import renderer from 'react-test-renderer';
import Feed from '../components/Feed';
import user from '../images/user.png';

const post = {
  id: 1,
  creator: 1,
  creatorName: 'Jack',
  corner: 'test_corner',
  cornerName: 'test_corner',
  type: 'post',
  pid: null,
  text: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit. Sed doloribus natus, nulla voluptates voluptate vitae amet qui consequuntur repellat repellendus sint! Non voluptas doloremque tempore ex perferendis labore illo totam.',
  pic: [{ user }],
  audio: null,
  video: null,
  createdAt: '2021-08-16T23:00:33.010+02:00',
  updatedAt: '2021-08-16T23:00:33.010+02:00',
  isFlagged: 1,
  isHidden: 0,
};

const Alldata = [
  {
    id: 1,
    creator: 1,
    creatorName: 'Jack',
    corner: 'test_corner',
    cornerName: 'test_corner',
    type: 'post',
    pid: null,
    text: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit. Sed doloribus natus, nulla voluptates voluptate vitae amet qui consequuntur repellat repellendus sint! Non voluptas doloremque tempore ex perferendis labore illo totam.',
    pic: [{ user }],
    audio: null,
    video: null,
    createdAt: '2021-08-16T23:00:33.010+02:00',
    updatedAt: '2021-08-16T23:00:33.010+02:00',
    isFlagged: 1,
    isHidden: 0,
  },
  {
    id: 2,
    creator: 1,
    creatorName: 'Amy',
    corner: 'test_corner',
    cornerName: 'test_corner',
    type: 'post',
    pid: null,
    text: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit. Sed doloribus natus, nulla voluptates voluptate vitae amet qui consequuntur repellat repellendus sint! Non voluptas doloremque tempore ex perferendis labore illo totam.',
    pic: [{ user }, { user }, { user }, { user }, { user }],
    audio: null,
    video: null,
    createdAt: '2021-08-16T23:00:33.010+02:00',
    updatedAt: '2021-08-16T23:00:33.010+02:00',
    isFlagged: 0,
    isHidden: 1,
  },
];

const noComment = [];
const comment = [
  {
    id: 3,
    creator: 1,
    creatorName: 'Jhon',
    corner: 'test_corner',
    cornerName: 'test_corner',
    type: 'comment',
    pid: 1,
    text: 'first test comment',
    pic: [],
    audio: null,
    video: null,
    createdAt: '2021-08-16T23:00:33.010+02:00',
    updatedAt: '2021-08-16T23:00:33.010+02:00',
    isFlagged: 0,
    isHidden: 0,
  },
  {
    id: 4,
    creator: 1,
    creatorName: 'Jhon',
    corner: 'test_corner',
    cornerName: 'test_corner',
    type: 'comment',
    pid: 1,
    text: 'second test comment',
    pic: [],
    audio: null,
    video: null,
    createdAt: '2021-08-16T23:00:33.010+02:00',
    updatedAt: '2021-08-16T23:00:33.010+02:00',
    isFlagged: 0,
    isHidden: 0,
  },
]

it('renders Feed correctly when user loggedin and is admin', () => {
  const tree = renderer.create(<Feed post={post} comments={noComment} allPosts={Alldata} isLoggedIn avatar={user} isAdmin cornerid='1' userName='test' userId='1' />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders Feed correctly when user not loggedin and is admin', () => {
  const tree = renderer.create(<Feed post={post} comments={noComment} allPosts={Alldata} isLoggedIn={false} avatar={user} isAdmin cornerid='1' userName='test' userId='1' />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders Feed correctly when user not loggedin and is not admin', () => {
  const tree = renderer.create(<Feed post={post} comments={noComment} allPosts={Alldata} isLoggedIn={false} avatar={user} isAdmin={false} cornerid='1' userName='test' userId='1' />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders Feed correctly when user loggedin and admin and has comment', () => {
  const tree = renderer.create(<Feed post={post} comments={noComment} allPosts={Alldata} isLoggedIn={false} avatar={user} isAdmin={false} cornerid='1' userName='test' userId='1' comments={comment} />).toJSON();
  expect(tree).toMatchSnapshot();
});