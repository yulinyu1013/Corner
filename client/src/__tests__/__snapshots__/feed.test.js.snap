// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Feed correctly when user loggedin and admin and has comment 1`] = `
<div
  className="feed-content"
>
  <div
    className="user-info"
  >
    <div
      className="left-info"
    >
      <img
        alt="user-pic"
        className="user-pic"
        src="logo.png"
      />
      <div
        className="title"
      >
        <p
          className="user-name"
        >
          Jack in test_corner
        </p>
      </div>
    </div>
    <div
      className="right-info"
    >
      <p
        className="time"
      >
        8/16/2021
      </p>
      <img
        alt="option"
        className="ant-dropdown-trigger option"
        onClick={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        src="more_vertical.svg"
      />
    </div>
  </div>
  <div
    className="post-content"
  >
    <p
      className="text-content"
    >
      Lorem ipsum dolor sit amet, consectetur adipisicing elit. Sed doloribus natus, nulla voluptates voluptate vitae amet qui consequuntur repellat repellendus sint! Non voluptas doloremque tempore ex perferendis labore illo totam.
    </p>
    <div
      className="feed-all-image"
    >
      <img
        alt="uploaded-pics"
        src={
          Array [
            "user.png",
          ]
        }
      />
    </div>
  </div>
  <div
    className="reaction"
  >
    <p>
      路
    </p>
    <p>
      2 comments
    </p>
  </div>
  <div
    className="comment-part"
  >
    <div
      className="comment-list"
    >
      <div
        className="comment-content"
      >
        <div
          className="single-comment"
        >
          <div
            className="top-content"
          >
            <img
              alt="commentUser"
              className="comment-pic"
              src="logo.png"
            />
            <div
              className="grey-box"
            >
              <p>
                Jhon
              </p>
              <p>
                first test comment
              </p>
            </div>
          </div>
          <div
            className="comment-operation"
          >
            <p>
              8/16/2021
            </p>
          </div>
        </div>
        <div
          className="single-comment"
        >
          <div
            className="top-content"
          >
            <img
              alt="commentUser"
              className="comment-pic"
              src="logo.png"
            />
            <div
              className="grey-box"
            >
              <p>
                Jhon
              </p>
              <p>
                second test comment
              </p>
            </div>
          </div>
          <div
            className="comment-operation"
          >
            <p>
              8/16/2021
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders Feed correctly when user loggedin and is admin 1`] = `
<div
  className="feed-content"
>
  <div
    className="user-info"
  >
    <div
      className="left-info"
    >
      <img
        alt="user-pic"
        className="user-pic"
        src="logo.png"
      />
      <div
        className="title"
      >
        <p
          className="user-name"
        >
          Jack in test_corner
        </p>
      </div>
    </div>
    <div
      className="right-info"
    >
      <p
        className="time"
      >
        8/16/2021
      </p>
      <img
        alt="option"
        className="ant-dropdown-trigger option"
        onClick={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        src="more_vertical.svg"
      />
    </div>
  </div>
  <div
    className="post-content"
  >
    <p
      className="text-content"
    >
      Lorem ipsum dolor sit amet, consectetur adipisicing elit. Sed doloribus natus, nulla voluptates voluptate vitae amet qui consequuntur repellat repellendus sint! Non voluptas doloremque tempore ex perferendis labore illo totam.
    </p>
    <div
      className="feed-all-image"
    >
      <img
        alt="uploaded-pics"
        src={
          Array [
            "user.png",
          ]
        }
      />
    </div>
  </div>
  <div
    className="reaction"
  >
    <p>
      路
    </p>
    <p>
      0 comments
    </p>
  </div>
  <div
    className="comment-part"
  >
    <div
      className="comment-list"
    >
      <div
        className="comment-content"
      />
      <div
        className="comment-box"
      >
        <img
          alt="user"
          className="user"
          src="user.png"
        />
        <div
          className="input input--singleLine"
          style={
            Object {
              "overflowY": "visible",
              "position": "relative",
            }
          }
        >
          <div
            className="input__control"
            style={
              Object {
                "border": "none",
                "display": "inline-block",
                "fontSize": 10,
              }
            }
          >
            <div
              className="input__highlighter input__highlighter--singleLine"
              style={
                Object {
                  "border": "1px solid transparent",
                  "boxSizing": "border-box",
                  "color": "transparent",
                  "overflow": "hidden",
                  "position": "relative",
                  "textAlign": "start",
                  "whiteSpace": "pre",
                  "width": "100%",
                  "wordWrap": null,
                }
              }
            >
               
            </div>
            <input
              className="input__input"
              onBlur={[Function]}
              onChange={[Function]}
              onCompositionEnd={[Function]}
              onCompositionStart={[Function]}
              onKeyDown={[Function]}
              onScroll={[Function]}
              onSelect={[Function]}
              placeholder="Write a comment..."
              style={
                Object {
                  "backgroundColor": "transparent",
                  "border": "1px solid #dfdfdf",
                  "borderRadius": 20,
                  "boxSizing": "border-box",
                  "display": "block",
                  "fontFamily": "inherit",
                  "fontSize": "inherit",
                  "left": 0,
                  "letterSpacing": "inherit",
                  "margin": 0,
                  "outline": "none",
                  "padding": 10,
                  "position": "absolute",
                  "top": 0,
                  "width": "100%",
                }
              }
              type="text"
              value=""
            />
          </div>
        </div>
        <button
          onClick={[Function]}
          type="submit"
        >
          POST
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders Feed correctly when user not loggedin and is admin 1`] = `
<div
  className="feed-content"
>
  <div
    className="user-info"
  >
    <div
      className="left-info"
    >
      <img
        alt="user-pic"
        className="user-pic"
        src="logo.png"
      />
      <div
        className="title"
      >
        <p
          className="user-name"
        >
          Jack in test_corner
        </p>
      </div>
    </div>
    <div
      className="right-info"
    >
      <p
        className="time"
      >
        8/16/2021
      </p>
      <img
        alt="option"
        className="ant-dropdown-trigger option"
        onClick={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        src="more_vertical.svg"
      />
    </div>
  </div>
  <div
    className="post-content"
  >
    <p
      className="text-content"
    >
      Lorem ipsum dolor sit amet, consectetur adipisicing elit. Sed doloribus natus, nulla voluptates voluptate vitae amet qui consequuntur repellat repellendus sint! Non voluptas doloremque tempore ex perferendis labore illo totam.
    </p>
    <div
      className="feed-all-image"
    >
      <img
        alt="uploaded-pics"
        src={
          Array [
            "user.png",
          ]
        }
      />
    </div>
  </div>
  <div
    className="reaction"
  >
    <p>
      路
    </p>
    <p>
      0 comments
    </p>
  </div>
  <div
    className="comment-part"
  >
    <div
      className="comment-list"
    >
      <div
        className="comment-content"
      />
    </div>
  </div>
</div>
`;

exports[`renders Feed correctly when user not loggedin and is not admin 1`] = `
<div
  className="feed-content"
>
  <div
    className="user-info"
  >
    <div
      className="left-info"
    >
      <img
        alt="user-pic"
        className="user-pic"
        src="logo.png"
      />
      <div
        className="title"
      >
        <p
          className="user-name"
        >
          Jack in test_corner
        </p>
      </div>
    </div>
    <div
      className="right-info"
    >
      <p
        className="time"
      >
        8/16/2021
      </p>
      <img
        alt="option"
        className="ant-dropdown-trigger option"
        onClick={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        src="more_vertical.svg"
      />
    </div>
  </div>
  <div
    className="post-content"
  >
    <p
      className="text-content"
    >
      Lorem ipsum dolor sit amet, consectetur adipisicing elit. Sed doloribus natus, nulla voluptates voluptate vitae amet qui consequuntur repellat repellendus sint! Non voluptas doloremque tempore ex perferendis labore illo totam.
    </p>
    <div
      className="feed-all-image"
    >
      <img
        alt="uploaded-pics"
        src={
          Array [
            "user.png",
          ]
        }
      />
    </div>
  </div>
  <div
    className="reaction"
  >
    <p>
      路
    </p>
    <p>
      0 comments
    </p>
  </div>
  <div
    className="comment-part"
  >
    <div
      className="comment-list"
    >
      <div
        className="comment-content"
      />
    </div>
  </div>
</div>
`;
